### PR TITLE
regional nudging using user provided nudging weights

### PR DESCRIPTION
**Pull request by xyuan**
regional nudging using user provided nudging weights file with the same gridset and vertical level

### CODE REVIEW

**Comment by PeterCaldwell** [comment]
@xyuan - is this PR ready to review?

**Comment by xyuan** [comment]
> @xyuan - is this PR ready to review?

yes, I tested ne4pg2, and ne30pg2 case, and both are working fine

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (AaronDonahue) [inline components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:48]: Is this left over from debugging?  If we do want to keep this message then I think we should use the atmosphere logger instead of printf.

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (AaronDonahue) [inline components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:161]: I don't think we can expect all nudging source files to use the hybrid coordinate system for the vertical pressure profile.  In fact, `p_levs` is used as the variable name for our vertical remapping because we pass explicit pressure levels for that.  

What is the goal with switching to `hybm` here?  I don't see this being used to build the pressure profile using the hybrid coordinate system anywhere, and wouldn't you need surface pressure as well?

If RRM nudging needs to use the hybrid coordinate system I would suggest adding another pressure type option to the nudging interface, in addition to `TIME_DEPENDENT_3D_PROFILE` and `STATIC_1D_VERTICAL_PROFILE`

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (AaronDonahue) [inline components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:356]: Question, do we apply weights by column or is it fully 3D?  If it's by column then we shouldn't need to vertically interpolate the weights as you suggested.

**Review by AaronDonahue — CHANGES_REQUESTED** [review]
Thanks for putting this together @xyuan!  I had a few questions about the PR, see below.

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (AaronDonahue) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:161]: Maybe a name like "TIME_DEPENDENT_3D_HYBRID" for one that uses horizontally varying surface pressure.  Of course you would also need to load `ps` from file.

Comment (xyuan) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:161]: sorry, this is also debug test, I should remove it

Comment (xyuan) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp]: since I don't have the plev data, so I tested it using hybm as an alternative option.

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (xyuan) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp]: yeah, let me remove it

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (AaronDonahue) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:349]: Ahh okay, that makes sense.  Then I would think we need to vertically interpolate the weights if we are already vertically interpolating the actual nudging data, to be consistent.  Also, one could imagine nudging data that is on a different vertical grid than the simulation, for example 72 vs. 128 levels.

Comment (xyuan) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:349]: sure, we can call the online mapping function to do the interpolation on vertical grid

File: components/eamxx/cime_config/namelist_defaults_scream.xml
Comment (PeterCaldwell) [inline components/eamxx/cime_config/namelist_defaults_scream.xml:234]: Would it make more sense to have "use_nudging_weights" be true or false rather than an integer? Is zero false and 1 true?

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (PeterCaldwell) [inline components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:107]: is dt the model timestep? I'm surprised it has integer type.

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (PeterCaldwell) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:349]: My vision was that the user would have to supply nudging weights on exactly the same grid as used for the physics parameterizations in the target simulation - no interpolation in horizontal, vertical or time (because the same weights would be used for all times). My logic is that it should be super-trivial to create a weights file for your target grid (just open an output file with a 3d output on that grid and the lat/lon dimensions, create a variable of the same size minus the time dimension, then loop over all the columns and assign a weight based on its lat/lon... Easier to create a new weights file than to add code for interpolation.

**Comment by PeterCaldwell** [comment]
Wouldn't it be simpler to have initialize_impl read in the nudging weights if m_use_weights is true and otherwise just initialize all the entries in nudging_weights_field to 1? Then instead of having an if/then for whether we have weights or not in apply_tendency you would always have weights. And wouldn't it be easier and equivalent to do base.update(tend,dtend*nudging_weights,1.0) in apply_tendency?

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (yuenxq) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:349]: @PeterCaldwell, sure, the code is working in the way you described at the moment

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (AaronDonahue) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:107]: Yes, the model timestep is passed as an integer to atmosphere processes.  I can't quite remember why, @bartgol may.

Comment (PeterCaldwell) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:107]: I think E3SM/CPL7 may be specifying dt as an integer (which saves a bit of memory), so it would be natural for us to do so as well. Still, I'm surprised we don't recast it to double. In particular, the ability to substep procs at runtime seems like it would naturally result in non-integer timesteps. For example, what if full-model dt is 150 sec and a user asks for 7 micro substeps (=21.43 sec micro dt)?

Comment (bartgol) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:107]: The timestep is _not_ an integer. It is passed by the cpl to the AD as an integer, but atm processes expect a double.

**Comment by bartgol** [comment]
> Wouldn't it be easier and equivalent to do base.update(tend,dtend*nudging_weights,1.0) in apply_weighted_tendency?

The `update` method does not support the 2nd argument (alpha) to be a field itself, it must be a scalar.

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (bartgol) [inline components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:73]: Why is this field exposed? It seems to me that this should be a field local to the atm proc.

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (bartgol) [inline components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:115]: Since `mPack` is just a pack of size 1, why can't we just use `Real` and avoid the confusion of packs?

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (bartgol) [inline components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:181]: Looks like `nudging_weights_ext` is never used? Can we prune it?

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (AaronDonahue) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:115]: Yeah, this is something that drove me crazy.  In short, the vertical interpolation function in EAMxx only works on Packs, so we cannot pass it an array of Reals.  I think the proper fix is to overhaul the vertical interpolation function to accept Real arrays.

Comment (bartgol) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:115]: Yeah, but there's no interpolation here...

Comment (AaronDonahue) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:115]: Ahh right, so we should change all for of these `get_view`'s to `Real`.

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (xyuan) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:73]: yes, this can be the local field, if we have the time-independent nudging weights, so it doesn't need to be the global.

**Comment by xyuan** [comment]
> > Wouldn't it be easier and equivalent to do base.update(tend,dtend*nudging_weights,1.0) in apply_weighted_tendency?
> 
> The `update` method does not support the 2nd argument (alpha) to be a field itself, it must be a scalar.

if the field update method support arguments of view type, it can remove the parallel_for loop. but it looks quite a lot work by doing so

File: components/eamxx/cime_config/namelist_defaults_scream.xml
Comment (xyuan) [reply components/eamxx/cime_config/namelist_defaults_scream.xml:234]: yeah, zero is false, and 1 is true, 
xml doesn't support _bool_, and _boolean_, see the error information below:
NameError: ERROR: Invalid/unsupported force type 'boolean', and 
NameError: ERROR: Invalid/unsupported force type 'bool'

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (xyuan) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:115]: ok, that looks less confusion here

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (PeterCaldwell) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:104]: Ok, so this should be fixed.

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (PeterCaldwell) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:344]: I think this issue can be resolved by just deleting the comment lines directly below.

**Comment by bartgol** [comment]
> > > Wouldn't it be easier and equivalent to do base.update(tend,dtend*nudging_weights,1.0) in apply_weighted_tendency?
> > 
> > 
> > The `update` method does not support the 2nd argument (alpha) to be a field itself, it must be a scalar.
> 
> if the field update method support arguments of view type, it can remove the parallel_for loop. but it looks quite a lot work by doing so

We can add that capability, but I think this would be the only use case. Since it's a small ||for, and adding it requires to also add testing, I'm inclined to not add the overload. If more places start to need a "weighted" update method, we can def think about it.

**Comment by xyuan** [comment]
@AaronDonahue the test case fails because the YAML nudging file is missing, could you tell where I can find these files, so I can generated the weight file based on your input

**Comment by xyuan** [comment]
add create_nudging_weights for the test, and it passes on ANL gpu machine using cuda/11.8.0

xyuan@gpu07:/gpfs/jlse-fs0/projects/climate/xyuan/scream.nudging/components/eamxx/build/tests/coupled/physics_only/shoc_p3_nudging> ctest
Test project /gpfs/jlse-fs0/projects/climate/xyuan/scream.nudging/components/eamxx/build/tests/coupled/physics_only/shoc_p3_nudging
    Start 1: create_vert_remap
1/5 Test #1: create_vert_remap ................   Passed    0.81 sec
    Start 2: create_nudging_weights
2/5 Test #2: create_nudging_weights ...........   Passed    0.57 sec
    Start 3: shoc_p3_source
3/5 Test #3: shoc_p3_source ...................   Passed    1.42 sec
    Start 4: shoc_p3_nudged
4/5 Test #4: shoc_p3_nudged ...................   Passed    1.50 sec
    Start 5: shoc_p3_nudged_remapped
5/5 Test #5: shoc_p3_nudged_remapped ..........   Passed    1.52 sec

100% tests passed, 0 tests failed out of 5

Total Test time (real) =   5.83 sec

File: components/eamxx/tests/coupled/physics_only/shoc_p3_nudging/CMakeLists.txt
Comment (bartgol) [inline components/eamxx/tests/coupled/physics_only/shoc_p3_nudging/CMakeLists.txt:34]: This won't work. Passing the same property twice causes the second instance to override the first one. To set multiple fixtures, you should do this:
```
  CreateUnitTestFromExec (....
     FIXTURES_REQUIRED "create_vertical_remap_file;create_nudging_weights_file")
```

**Review by bartgol — CHANGES_REQUESTED** [review]

File: components/eamxx/tests/coupled/physics_only/shoc_p3_nudging/CMakeLists.txt
Comment (xyuan) [reply components/eamxx/tests/coupled/physics_only/shoc_p3_nudging/CMakeLists.txt:34]: but I have this error, when I use FIXTURES_REQUIRED "create_vertical_remap_file;create_nudging_weights_file")

CMake Error at /gpfs/jlse-fs0/projects/climate/xyuan/scream.nudging/externals/ekat/cmake/EkatCreateUnitTest.cmake:375 (set_tests_properties):
  set_tests_properties called with incorrect number of arguments.
Call Stack (most recent call first):
  cmake/ScreamUtils.cmake:130 (EkatCreateUnitTestFromExec)
  tests/coupled/physics_only/shoc_p3_nudging/CMakeLists.txt:30 (CreateUnitTestFromExec)

Comment (bartgol) [reply components/eamxx/tests/coupled/physics_only/shoc_p3_nudging/CMakeLists.txt:34]: Ah, yes, we have some issue with passing some properties that are lists. I tried this with `add_test`, but not with the Ekat utility. Fortunately, we have a special handling for FIXTURES...but unfortunately there is a bug in it. If you can hold this PR for a bit more, I can fix it in Ekat, then update ekat in scream repo?

Comment (xyuan) [reply components/eamxx/tests/coupled/physics_only/shoc_p3_nudging/CMakeLists.txt:34]: sure, please go ahead, it is possibly the reason that the cpu test failed, it shows _"Failed test depencies: shoc_p3_source"_

Comment (bartgol) [reply components/eamxx/tests/coupled/physics_only/shoc_p3_nudging/CMakeLists.txt:34]: The PR was merged in ekat. #2586 is undergoing testing right now. Once merged, we can re-test this, and should hopefully pass.

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp
Comment (bartgol) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.cpp:104]: @xyuan we still need to fix dt here. The method should accept a Real, not an integer, to avoid silent truncation by the compiler.

File: components/eamxx/cime_config/namelist_defaults_scream.xml
Comment (bartgol) [reply components/eamxx/cime_config/namelist_defaults_scream.xml:234]: Ah, we are using `logical` in our XML settings. Can you try with that?

**Review by bartgol — CHANGES_REQUESTED** [review]

File: components/eamxx/tests/coupled/physics_only/shoc_p3_nudging/create_vert_remap_and_weights.cpp
Comment (bartgol) [inline components/eamxx/tests/coupled/physics_only/shoc_p3_nudging/create_vert_remap_and_weights.cpp:82]: These two calls have the same parameters, so I'm guessing it will create the same output. Do we really need two files that are identicals?

**Review by bartgol — DISMISSED** [review]

**Comment by bartgol** [comment]
Github is blocking the PR merge b/c Aaaron requested changes, so it waits for his approval. I think he may be out today. Since his comments have been addressed, I have removed him as a reviewer, so that the integration can continue. I am not sure if gh will still prevent the PR from merging even if he's no longer a reviewer. Maybe. If so, I can force-merge the PR.

File: components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.hpp
Comment (AaronDonahue) [reply components/eamxx/src/physics/nudging/eamxx_nudging_process_interface.hpp:121]: It looks like this is still an `int`.  Should it b e a `bool`?

**Review by AaronDonahue — CHANGES_REQUESTED** [review]

**Review by bartgol — APPROVED** [review]

File: components/eamxx/cime_config/namelist_defaults_scream.xml
Comment (bartgol) [reply components/eamxx/cime_config/namelist_defaults_scream.xml:234]: Another option would be to actually set a default... Empty parameters kind of make sense for lists/strings, while for other types, leaving them empty makes the interpretation somewhat arbitrary...

Comment (bartgol) [reply components/eamxx/cime_config/namelist_defaults_scream.xml:234]: I now see that that's exactly what you did in the PR you linked. :)

**Pull request merged**
